### DISCUSSION TITLE
dhcp relay issue over ipsec/ikev2 tunnel

### DISCUSSION THREAD

**Discussion by thom24** [strongswan/strongswan#675] [Q&A] [answered]
Hello there,

I try to setup the following configuration:
- an IPSEC/IKEv2 tunnel between my router and remote server.
- my router has a wifi hotspot and forward incoming wifi traffic through the ipsec tunnel.
- the gateway of wifi client is my router (10.132.80.254)
- the wifi client gets its ip address from a remote dhcp server located behind the remote server

![image](https://user-images.githubusercontent.com/24258848/137921139-70e0a7b4-6aa4-460c-bad8-208ee63a3387.png)

My router has one default route which is the lte interface.

```

$ cat /etc/ipsec.conf

conn %default
        ikelifetime=3600s
        keylife=60m
        rekeymargin=3m
        keyingtries=%forever
        keyexchange=ikev2
        authby=secret
        
conn tunnel
        auto=start
        left=%any
        leftid=XXX
        leftsubnet=10.132.80.224/27
        rightsubnet=0.0.0.0/0
        leftauth=secret
        right=63.156.152.100
        rightid=YYY
        rightauth=secret
        ike=aes256-sha256-prfsha256-modp1536
        esp=aes256-sha256-prfsha256-modp1536

$ cat /etc/strongswan.conf
charon {
        load_modular = yes
        install_routes=no
        install_virtual_ip=no
        plugins {
                include strongswan.d/charon/*.conf
        }
}

include strongswan.d/*.conf

include /var/ipsec/strongswan.conf
```
The tunnel is running well:
```
Connections:
     tunnel:  %any...63.156.152.100  IKEv2
     tunnel:   local:  [XXX] uses pre-shared key authentication
     tunnel:   remote: [YYY] uses pre-shared key authentication
     tunnel:   child:  10.132.80.224/27 === 0.0.0.0/0 TUNNEL
Security Associations (1 up, 0 connecting):
     tunnel[1]: ESTABLISHED 48 minutes ago, 10.12.241.175[XXX]...63.156.152.100[YYY]
     tunnel[1]: IKEv2 SPIs: 98593e0917c7e646_i* 8f781c53405c0c57_r, pre-shared key reauthentication in 6 minutes
     tunnel[1]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536
     tunnel{1}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: cf9f8d40_i 844093b6_o
     tunnel{1}:  AES_CBC_128/HMAC_SHA1_96, 4800 bytes_i (50 pkts, 52s ago), 4608 bytes_o (48 pkts, 52s ago), rekeying in 6 minutes
     tunnel{1}:   10.132.80.224/27 === 0.0.0.0/0
```
The config seems to be valid, as if a wifi client with an ip address set by hands can ping the dhcp server.
So packets are well routed through the tunnel.

But i need dhcp support.
To implement the dhcp relay, i set up dnmasq with the following option
```
dhcp-relay=10.132.80.254,172.15.153.234
```
Dnsmasq receives well the packet and forward it to the dhcp server.
When i run tcpdump on my router, i can see the dhcp reply packet (only on the lte interface), which is very strange. The dnsmasq does not see the reply, so it cannot be forward it to the client.

I also saw dhcp options in charon, i tried to use it without success.
But i don't know if it is the solution for my case.

Thanks for your help.

Thomas

**Comment by tobiasbrunner** [comment, marked as answer]
> I also saw dhcp options in charon, i tried to use it without success.
> But i don't know if it is the solution for my case.

No, that feature is used to assign virtual IPs to the VPN client (i.e. your router), not to hosts behind a VPN client.

> When i run tcpdump on my router, i can see the dhcp reply packet (only on the lte interface), which is very strange. The dnsmasq does not see the reply, so it cannot be forward it to the client.

Maybe it's listening on the wrong interface(s) or there are some restrictive firewall rules.

**Reply by thom24** [reply]
strongswan send decrypted dhcp packet only on lte interface whereas destination ip of the dhcp offer packet is 10.132.80.254 (wifi ip address).
And dhcp relay does not listen on lte interface as the lte interface is rawip

**Reply by tobiasbrunner** [reply]
Not sure what you mean with "rawip". But as far as I understand, dnsmasq should listen on all interfaces, by default (i.e. if you don't configure any interface or listen address).

Anyway, it's not strongSwan, but the Linux kernel who does that. And to Linux it doesn't matter on which interface a local IP address is, or on which interface a packet with such a destination address is received (the address could even be on `lo` and packets addressed to it would be accepted on any interface). Furthermore, unless you use [XFRM or VTI interfaces](https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN), the interface associated with a decrypted packet is probably always the same as that on which the ESP packet was received. So it might be best to not bind dnsmasq to any particular interface/address and use firewall rules with policy matching or e.g. marks to prevent that it processes packets that weren't received via IPsec from the LTE interface.

**Reply by thom24** [reply]
Ok, i understand.
The lte modem is working in raw-ip mode.
It means that my lte interface does not have mac address
```
root@gbNbMBtWvg:~# ifconfig ethLTE
ethLTE    Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  
          inet addr:10.250.16.119  P-t-P:10.250.16.119  Mask:255.255.255.240
          inet6 addr: fe80::4e25:6759:cfb4:e709/64 Scope:Link
          UP POINTOPOINT RUNNING NOARP MULTICAST  MTU:1500  Metric:1
          RX packets:1171 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1508 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:107928 (105.3 KiB)  TX bytes:165282 (161.4 KiB)
```

So dnsmasq (or dhcrelay) does not want to listen on this interface.
So i guess  I have to forward packets to an other interface (an interface that dnsmasq can listen on) using firewall rules.
I gonna try this.
